File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        ans = Integer.MIN_VALUE;
        Solve(root);
        return ans;
    }
    static int ans;
    public static int Solve(TreeNode root){
        if(root ==  null){
            return 0;
        }
        int left = Solve(root.left);
        int right = Solve(root.right);
        int curr1 = root.val + left + right;
        int curr2 = root.val+ left;
        int curr3 = root.val+ right;
        int curr4 = root.val;
        ans = Math.max(ans,Math.max(curr1,Math.max(curr2,Math.max(curr3,curr4))));
        return Math.max(root.val, root.val+ Math.max(left,right));
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum;
    int maxPathSum(TreeNode* root) {
        maxSum = INT_MIN;
        Solve(root);
        return maxSum;
    }
    int Solve(TreeNode* root){
        if(root==nullptr){
            return 0;
        }
        int left = Solve(root->left);
        int right = Solve(root->right);
        int curr = root->val + max(0,left)+ max(0,right);
        maxSum = max(maxSum,curr);

        return max(0,max(left,right))+ root->val;
    }
};